### PR TITLE
Make sequtils.zip return seq of anonymous tuples

### PR DESCRIPTION
**Pull request by kaushalmodi**
Earlier the tuples had named fields "a" and "b" and that made it
difficult to assign the zip returned seqs to other vars which expected
seqs of tuples with field names other than "a" and "b".

### CODE REVIEW

**Comment by kaushalmodi** [comment]
@krux02 That's a good idea. I'll get such a test in by tomorrow.

**Comment by krux02** [comment]
I just want to mention that I was once bothered by the field names in the tuple as well. So thumbs up for this PR.

**Comment by kaushalmodi** [comment]
Thanks :) I have updated the tests.

File: lib/pure/collections/sequtils.nim [hunk 1]
```diff
@@ -1070,18 +1067,18 @@ when isMainModule:
       zip1 = zip(short, long)
       zip2 = zip(short, words)
       zip3 = zip(ashort, along)
-      zip4 = zip(ashort, awords)
-      zip5 = zip(ashort, words)
+      zipXy: seq[tuple[x: int, y: string]] = zip(ashort, awords)
+      zipMn: seq[tuple[m: int, n: string]] = zip(ashort, words)
```
Comment (krux02) [inline lib/pure/collections/sequtils.nim]: I see that you introduce the test here. A comment of why you have this specific assignment is important. I often encounter tests that do things and I never know if something is intentional or not. I often wonder the intention of a test when I read it. Unlike other code, the context of a test does not provide the intention.

Apart from that I think this PR is fine and can be merged.

Comment (kaushalmodi) [reply lib/pure/collections/sequtils.nim]: Done!

**Comment by mratsim** [comment]
So anonymous tuples can be assigned to other anonymous tuples or to named tuples (i.e. structural typing) but named tuples can only be assigned to other named tuples (i.e. "string"/"identifier" typing).

I think this part is underspecified. Also while we tuple destructuring works for both anonymous and named tuple, maybe a function or magic that anonymize named tuples is needed for function call chaining like the original case?

**Comment by kaushalmodi** [comment]
> but named tuples can only be assigned to other named tuples (i.e. "string"/"identifier" typing)

No, named tuples can be assigned to named tuples with the exact name fields, or also to any anonymous tuple.

**Comment by Varriount** [comment]
Does this break backwards compatibility at all?

**Comment by kaushalmodi** [comment]
@Varriount Yes, slightly, if someone is relying on inferred typing.

E.g. this will break:

```
let foo = zip(someSeq1, someSeq2)
echo foo.a
```

But this will not break:

```
let foo: seq[tuple[a:int, b:int]] = zip(someSeq1, someSeq2) # of course the types of a and b fields are assumed here
echo foo.a
```

**Comment by disruptek** [comment]
Why would you not want names?  They are the only way to confirm that the order and types of the tuple you are destructuring match your expectation.

That's the whole point of my `foreach` macro, BTW.

**Comment by disruptek** [comment]
Ignore me.  Confused this for a different issue.  :laughing:

**Comment by krux02** [comment]
Good point about the breaking behavior. We do have a backwards compatibility claim. But I am still positive for this change, since I really do like it. You should introduce a backwards compatibility branch. This is how you do it:

```nim
when (NimMajor, NimMinor) <= (1, 0) # 1.0.x
  # old declaration here
else:
  # new declration here
```

Of course you also need to have a test for the old version (reintroduce the broken tests into another file that is compiled with ``--useVersion:1.0``. Sorry for this inconvenience.

**Comment by kaushalmodi** [comment]
OK, I'll get to this later today. 

I have a question though..

Does the CI run all the tests with and without `--useVersion:1.0`.

**Comment by Araq** [comment]
> Does the CI run all the tests with and without --useVersion:1.0.

No and it's worse than that, right now ``--useVersion:1.0`` doesn't do anything (yet!).

**Comment by kaushalmodi** [comment]
Also, looks like `--useVersion:1.0` would apply to the whole core? So if someone wants old version of `zip`, they would need to take the old versions of everything?

**Comment by Araq** [comment]
Yes, it applies to the whole core.

**Comment by kaushalmodi** [comment]
@krux02 , @Araq : Please review the new commit added to make `zip` backwards compatible: https://github.com/nim-lang/Nim/pull/12575/commits/576993fc0b9bbaea6d30ba37d5bc149ac97d7939

**Comment by kaushalmodi** [comment]
I have redone the code change to make `zip` backward compatible using the tip from @PMunch (https://play.nim-lang.org/#ix=20MO).

/cc @narimiran @PMunch  Can you please review https://github.com/nim-lang/Nim/pull/12575/commits/c0b597c24a6d43420ce59ae98e378b6bfc28b073 ?

**Pull request merged**